### PR TITLE
Use grouped validator instead of categories

### PR DESCRIPTION
**Pull request by tonyzhao1**
Removes validator categories and instead creates a GroupedValidator that takes a list of sub-validators and handles merging. This moves the logic closer to where it's used (in the validator itself) and removes some of the case handling that was needed with categories. 

Also splits the iOS validator into a base iOS validator, and a CocoaPods-only validator.

### CODE REVIEW

**Comment by tonyzhao1** [comment]
@tvolkert @mehmetf I feel silly undoing so many of my changes but this seems so much cleaner and so much more intuitive.

File: packages/flutter_tools/lib/src/doctor.dart [hunk 1]
```diff
@@ -51,7 +51,7 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
         _validators.add(androidValidator);
 
       if (iosWorkflow.appliesToHostPlatform)
-        _validators.add(iosValidator);
+        _validators.add(new GroupedValidator(<DoctorValidator>[iosValidator, cocoapodsValidator]));
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/doctor.dart]: Will this affect the presentation of the iOS validation to public users?

**Review by mehmetf — COMMENTED** [review]
Much cleaner, thanks!

Could you post the output of doctor with these changes on mac and linux? That will also answer one of Todd's questions.

File: packages/flutter_tools/lib/src/doctor.dart [hunk 2]
```diff
@@ -180,42 +159,18 @@ class Doctor {
...
-        }
+      try {
+        await validatorTask.result;
+      } catch (exception) {
+        status.cancel();
+        rethrow;
       }
       status.stop();
 
+      final ValidationResult result = await validatorTask.result;
```
[the remark this answers is not in the document]
Comment (tonyzhao1) [reply packages/flutter_tools/lib/src/doctor.dart]: Done

File: packages/flutter_tools/lib/src/doctor.dart [hunk 3]
```diff
@@ -318,32 +241,69 @@ enum ValidationType {
...
-  const DoctorValidator(this.title, [this.category = ValidatorCategory.other]);
+  const DoctorValidator(this.title);
 
   final String title;
-  final ValidatorCategory category;
 
   Future<ValidationResult> validate();
 }
 
+class GroupedValidator extends DoctorValidator {
```
[the remark this answers is not in the document]
Comment (tonyzhao1) [reply packages/flutter_tools/lib/src/doctor.dart]: Done

File: packages/flutter_tools/lib/src/doctor.dart [hunk 4]
```diff
@@ -318,32 +241,69 @@ enum ValidationType {
...
+    for (DoctorValidator validator in subValidators) {
+      tasks.add(new ValidatorTask(validator, validator.validate()));
+    }
+
+    final List<ValidationResult> results = <ValidationResult>[];
+    for (ValidatorTask subValidator in tasks) {
+      try {
+        results.add(await subValidator.result);
+      } catch (exception) {
+        rethrow;
```
[the remark this answers is not in the document]
Comment (tonyzhao1) [reply packages/flutter_tools/lib/src/doctor.dart]: Done

File: packages/flutter_tools/lib/src/doctor.dart [hunk 5]
```diff
@@ -318,32 +241,69 @@ enum ValidationType {
...
+      }
+    }
+    return _mergeValidationResults(results);
+  }
+
+  ValidationResult _mergeValidationResults(List<ValidationResult> results) {
+    ValidationType mergedType = results[0].type;
```
[the remark this answers is not in the document]
Comment (tonyzhao1) [reply packages/flutter_tools/lib/src/doctor.dart]: Done

File: packages/flutter_tools/lib/src/doctor.dart (hunk 1 above)
Comment (tonyzhao1) [reply packages/flutter_tools/lib/src/doctor.dart]: It shouldn't. Pasted doctor output in a comment below

**Comment by tonyzhao1** [comment]
Running doctor -v on my incompletely configured environments.

Linux: 

[✓] Flutter (Channel master, v0.5.9-pre.38, on Linux, locale en_US.UTF-8)
    • Flutter version 0.5.9-pre.38 at /usr/local/google/home/tonyzhao/flutter
    • Framework revision f4ec5fbf0c (2 hours ago), 2018-09-17 11:34:44 -0700
    • Engine revision 1b2a2075a9
    • Dart version 2.1.0-dev.4.0.flutter-bb56d4592f

[!] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /usr/local/google/home/tonyzhao/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /opt/android-studio-with-blaze-3.0/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Android Studio (version 3.0)
    • Android Studio at /opt/android-studio-with-blaze-3.0
    • Flutter plugin version 23.2.1
    • Dart plugin version 171.4424
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)

[✓] IntelliJ IDEA Community Edition (version 2018.1)
    • IntelliJ at /opt/intellij-ce-2018.1
    • Flutter plugin version 28.0.2
    • Dart plugin version 181.5616

[!] Connected devices
    ! No devices available

! Doctor found issues in 2 categories.

Mac: 

[✓] Flutter (Channel master, v0.5.9-pre.38, on Mac OS X 10.13.3 17D102, locale en-US)
    • Flutter version 0.5.9-pre.38 at /Users/tonyzhao/Documents/GitHub/flutter
    • Framework revision f4ec5fbf0c (2 hours ago), 2018-09-17 11:34:44 -0700
    • Engine revision 1b2a2075a9
    • Dart version 2.1.0-dev.4.0.flutter-bb56d4592f

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set $ANDROID_HOME to that location.

[!] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 2.0.0
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without resolving iOS dependencies with CocoaPods, plugins will not work on iOS.
        For more info, see https://flutter.io/platform-plugins
      To install:
        brew install cocoapods
        pod setup

[✗] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[!] Connected devices
    ! No devices available

! Doctor found issues in 4 categories.

**Review by mehmetf — APPROVED** [review]
This LGTM

**Pull request merged**